modules/frontend/pipeline/trace_query_filter.go

```diff
@@ -0,0 +1,105 @@
+package pipeline
+
+import (
+	"fmt"
+	"github.com/grafana/tempo/pkg/api"
+	"io"
+	"net/http"
+	"regexp"
+	"strings"
+	"sync"
+)
+
+type traceQueryFilterWare struct {
+	next    http.RoundTripper
+	filters []*regexp.Regexp
+	parseFn func(r *http.Request) string
+}
+
+func NewTraceQueryFilterWare(next http.RoundTripper) http.RoundTripper {
+	return &traceQueryFilterWare{
+		next: next,
+	}
+}
+
+func NewTraceQueryFilterWareWithDenyList(denyList []string, parseFunc func(r *http.Request) string) Middleware {
+	var filter []*regexp.Regexp
+	for i := range denyList {
+		exp, err := regexp.Compile(denyList[i])
+		if err == nil {
+			filter = append(filter, exp)
+		}
+	}
+
+	return MiddlewareFunc(func(next http.RoundTripper) http.RoundTripper {
+		return traceQueryFilterWare{
+			next:    next,
+			filters: filter,
+			parseFn: parseFunc,
+		}
+	})
+}
+
+func (c traceQueryFilterWare) RoundTrip(req *http.Request) (*http.Response, error) {
+	if c.filters == nil || len(c.filters) == 0 {
+		return c.next.RoundTrip(req)
+	}
+
+	query := c.parseFn(req)
+
+	if len(query) == 0 || query == "" {
+		return c.next.RoundTrip(req)
+	}
+
+	match := make(chan bool, len(c.filters))
+	wg := sync.WaitGroup{}
+	for range c.filters {
```

Comment:
no need for concurrency. let's keep this simple and just check them one at a time inline